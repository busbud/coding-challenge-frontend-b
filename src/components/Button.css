.button {
  padding: 0.25rem 1rem;
  color: white;
  font-size: 2rem;
  border-radius: 8px;
  display: inline-block;
  text-align: center;
  transition: background 0.5s;
}
.button:hover {
  cursor: pointer;
}
.button--primary {
  background-color: orange;
}
.button--primary:hover {
  background-color: #d28a06;
}
.button--secondary {
  background-color: skyblue;
}
