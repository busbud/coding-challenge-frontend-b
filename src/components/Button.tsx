const Button = ({
  className = 'button',
  defaultText = 'Click me',
  disabled = false,
  onClick = () => {},
  dataTestId = '',
}: Props) => {
  return (
    <button
      className={`${className} ${disabled ? 'button__disabled' : ''}`}
      disabled={disabled}
      onClick={onClick}
      type="button"
      data-testid={dataTestId}
    >
      {defaultText}
    </button>
  );
};

export default Button;

interface Props {
  className?: string;
  disabled?: boolean;
  defaultText: string;
  onClick: () => void;
  dataTestId: string;
}
