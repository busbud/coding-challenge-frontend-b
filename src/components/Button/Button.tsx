import React from 'react';
import styled from 'styled-components';

interface Props {
  onClickAction: (payload: string) => void;
  children: React.ReactNode;
}

const StyledButton = styled.button`
  background-color: #161616;
  border-radius: 5px;
  color: white;
  font-size: 0.8rem;
  padding: 10px;
`;

export default function Button({ onClickAction, children }: Props) {
  return (
    <StyledButton type="button" onClick={() => onClickAction('todo')}>
      {children}
    </StyledButton>
  );
}
