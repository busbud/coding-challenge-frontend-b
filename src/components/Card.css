.card {
  padding: 2rem;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.24);
}
