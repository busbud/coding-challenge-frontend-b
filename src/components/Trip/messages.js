import { defineMessages } from 'react-intl';

export default defineMessages({
  action_booking: {
    id: 'components.trip.action.booking',
    defaultMessage: 'Book now',
  },
  hours: {
    id: 'components.trip.hours',
    defaultMessage: '{amount}H ',
  },
  minutes: {
    id: 'components.trip.minutes',
    defaultMessage: '{amount}m',
  },
});
