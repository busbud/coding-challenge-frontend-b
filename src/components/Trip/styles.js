export default theme => ({
  departure: {
    textAlign: 'left',
  },
  duration: {
    textAlign: 'center',
  },
  arrival: {
    textAlign: 'right',
  },
  actions: {
    display: 'flex',
  },
  expand: {
    transform: 'rotate(0deg)',
    transition: theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    }),
    marginLeft: 'auto',
  },
  expandOpen: {
    transform: 'rotate(180deg)',
  },
});
