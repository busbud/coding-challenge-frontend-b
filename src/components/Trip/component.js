/* eslint-disable react/style-prop-object */
import React from 'react';
import { object, string, number, bool, func } from 'prop-types';
import { FormattedMessage, FormattedDate, FormattedTime, FormattedNumber } from 'react-intl';
import Grid from 'material-ui/Grid';
import Hidden from 'material-ui/Hidden';
import Collapse from 'material-ui/transitions/Collapse';
import IconButton from 'material-ui/IconButton';
import Card, { CardActions, CardContent } from 'material-ui/Card';
import Button from 'material-ui/Button';
import Typography from 'material-ui/Typography';
import DirectionsBusIcon from 'material-ui-icons/DirectionsBus';
import ExpandMoreIcon from 'material-ui-icons/ExpandMore';
import { getHours, getMinutes, formattedAddress } from '../../utils/helpers';
import messages from './messages';

const propTypes = {
  classes: object.isRequired,
  departure_time: string.isRequired,
  arrival_time: string.isRequired,
  duration: number.isRequired,
  origin_location: object.isRequired,
  destination_location: object.isRequired,
  prices: object.isRequired,
  links: object.isRequired,
  isExpanded: bool.isRequired,
  toggleExpansion: func.isRequired,
};

const Component = ({
  classes,
  departure_time,
  arrival_time,
  duration,
  origin_location,
  destination_location,
  prices,
  links,
  isExpanded,
  toggleExpansion,
}) => {
  const departureTime = new Date(departure_time);
  const arrivalTime = new Date(arrival_time);
  const durationHours = getHours(duration);
  const durationMinutes = getMinutes(duration);

  return (
    <Grid item md={6} xs={12}>
      <Card>
        <CardContent>
          <Grid container justify="center" alignItems="center">
            <Grid className={classes.departure} item md={5} xs={6}>
              <Typography gutterBottom>
                <FormattedDate value={departureTime} month="long" day="2-digit" />
              </Typography>

              <Typography variant="display1">
                <FormattedTime value={departureTime} />
              </Typography>
            </Grid>
            <Hidden mdDown>
              <Grid className={classes.duration} item xs={2}>
                <DirectionsBusIcon color="action" />
                <Typography className={classes.duration}>
                  {durationHours && (
                    <FormattedMessage
                      className="duration__hours"
                      values={{ amount: durationHours }}
                      {...messages.hours}
                    />
                  )}
                  {durationMinutes && (
                    <FormattedMessage
                      className="duration__minutes"
                      values={{ amount: durationMinutes }}
                      {...messages.minutes}
                    />
                  )}
                </Typography>
              </Grid>
            </Hidden>
            <Grid className={classes.arrival} item md={5} xs={6}>
              <Typography gutterBottom>
                <FormattedDate value={arrivalTime} month="long" day="2-digit" />
              </Typography>

              <Typography variant="display1">
                <FormattedTime value={arrivalTime} />
              </Typography>
            </Grid>
          </Grid>
        </CardContent>

        <Collapse in={isExpanded} timeout="auto" unmountOnExit>
          <CardContent>
            <Grid container justify="space-between" alignItems="center">
              <Grid className={classes.departure} item xs={5}>
                <Typography variant="title">{origin_location.name}</Typography>
                <Typography>{formattedAddress(origin_location)}</Typography>
              </Grid>
              <Grid className={classes.arrival} item xs={5}>
                <Typography variant="title">{destination_location.name}</Typography>
                <Typography>{formattedAddress(destination_location)}</Typography>
              </Grid>
            </Grid>
          </CardContent>
        </Collapse>

        <CardActions className={classes.actions} disableActionSpacing>
          <Typography variant="button">
            <FormattedNumber value={prices.total / 100} style="currency" currency={prices.currency} />
          </Typography>
          <Button href={links.deeplink} target="_blank" color="primary">
            <FormattedMessage {...messages.action_booking} />
          </Button>
          <IconButton
            className={`${classes.expand} ${isExpanded ? classes.expandOpen : ''}`}
            onClick={toggleExpansion}
            aria-expanded={isExpanded}
            aria-label="Show more"
          >
            <ExpandMoreIcon />
          </IconButton>
        </CardActions>
      </Card>
    </Grid>
  );
};

Component.propTypes = propTypes;

export default Component;
