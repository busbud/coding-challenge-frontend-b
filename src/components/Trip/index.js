import { compose, withStateHandlers } from 'recompose';
import { withStyles } from 'material-ui/styles';
import styles from './styles';
import component from './component';

export const enhance = compose(
  withStateHandlers(
    ({ initialIsExpanded = false }) => ({
      isExpanded: initialIsExpanded,
    }),
    {
      toggleExpansion: ({ isExpanded }) => value => ({
        isExpanded: !isExpanded,
      }),
    },
  ),
  withStyles(styles),
);

export default enhance(component);
