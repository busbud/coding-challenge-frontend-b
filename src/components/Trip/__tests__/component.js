import React from 'react';
import { shallow } from 'enzyme';
import { FormattedMessage, FormattedDate, FormattedTime, FormattedNumber } from 'react-intl';
import Grid from 'material-ui/Grid';
import Hidden from 'material-ui/Hidden';
import Collapse from 'material-ui/transitions/Collapse';
import IconButton from 'material-ui/IconButton';
import Card, { CardActions, CardContent } from 'material-ui/Card';
import Button from 'material-ui/Button';
import Typography from 'material-ui/Typography';
import DirectionsBusIcon from 'material-ui-icons/DirectionsBus';
import ExpandMoreIcon from 'material-ui-icons/ExpandMore';
import { getHours, getMinutes, formattedAddress } from '../../../utils/helpers';
import messages from '../messages';
import Trip from '../component';

const props = {
  classes: {
    departure: 'Component-departure-274',
    duration: 'Component-duration-275',
    arrival: 'Component-arrival-276',
    actions: 'Component-actions-277',
    expand: 'Component-expand-278',
    expandOpen: 'Component-expandOpen-279',
  },
  arrival_timezone: 'America/Montreal',
  available_seats: 50,
  bus: null,
  busbud_departure_id: 'c1c8c842',
  class: 'Economy',
  class_name: '-',
  fare_name: null,
  deeplink: null,
  departure_timezone: 'America/New_York',
  destination_location_id: 1938,
  duration: 499,
  links: {
    deeplink:
      'https://www.busbud.com/en/deeplink/dr5reg/f25dvk/N2Y5ODU1MTpjMWM4Yzg0Mg?outbound_date=2018-08-07&return_date&adults=1&children=0&seniors=0&child_ages=&senior_ages=&discount_code&utm_source={utm_source}&utm_campaign={utm_campaign}&utm_medium={utm_medium}',
  },
  operator_id: 'bfc27cd5-44ca-49c1-8d00-0f2bc00c58c0',
  origin_location_id: 1942,
  source_id: 155,
  ticket_types: ['print', 'claim'],
  departure_time: '2018-08-07T00:01:00',
  arrival_time: '2018-08-07T08:20:00',
  fetched_at: '2018-03-11T17:55:24.569Z',
  prices: {
    currency: 'USD',
    total: 7200,
    categories: {
      adult: 7200,
      child: 7200,
      senior: 7200,
    },
    discount: 0,
    roundtrip_min: null,
    roundtrip_total: null,
    discounted: null,
    breakdown: {
      base: 7200,
      fees: 0,
      taxes: 0,
      discount: 0,
    },
  },
  trip_stops: null,
  addons: null,
  details: {},
  operator: {
    id: 'bfc27cd5-44ca-49c1-8d00-0f2bc00c58c0',
    source_id: 155,
    profile_id: 580,
    name: 'Greyhound',
    url: null,
    logo_url: 'https://busbud.imgix.net/operator-logos/greyhound.png?h={height}&w={width}&auto=format&fit=fill&bg=0FFF',
    display_name: 'Greyhound',
  },
  destination_location: {
    id: 1938,
    city_id: '375dd587-9001-acbd-84a4-683dedfb933e',
    name: "Gare d'autocars de Montréal",
    address: ['1717 Rue Berri', 'Montréal, QC H2L 4E9', 'Canada'],
    type: 'bus_station',
    lat: 45.516235,
    lon: -73.562668,
    geohash: 'f25dyjcr3',
  },
  origin_location: {
    id: 1942,
    city_id: '375dd587-9001-acbd-84a4-683deda84183',
    name: 'Port Authority Bus Terminal',
    address: ['619-623 8th Ave', 'New York, NY 10018', 'USA'],
    type: 'bus_station',
    lat: 40.756252,
    lon: -73.990684,
    geohash: 'dr5ru73wk',
  },
  isExpanded: false,
  toggleExpansion: jest.fn(),
};

describe('component | Jumbotron | component', () => {
  describe('render', () => {
    it('should render without throwing and error', () => {
      // given
      const wrapper = shallow(<Trip {...props} />);

      // then
      expect(wrapper).toBeDefined();
    });
    it('should match snapshot', () => {
      // given
      const wrapper = shallow(<Trip {...props} />);

      // then
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('CardContent', () => {
    describe('Departure', () => {
      it('should have proper className props', () => {
        // given
        const wrapper = shallow(<Trip {...props} />);

        const { className } = wrapper
          .find(CardContent)
          .at(0)
          .find(Grid)
          .at(1)
          .props();

        expect(className).toBe(props.classes.departure);
      });
      describe('FormattedDate', () => {
        it('should contains proper value', () => {
          // given
          const wrapper = shallow(<Trip {...props} />);

          const { value } = wrapper
            .find(CardContent)
            .at(0)
            .find(Grid)
            .at(1)
            .find(FormattedDate)
            .props();

          // then
          const expectedValue = new Date(props.departure_time);
          expect(value).toEqual(expectedValue);
        });
      });
      describe('FormattedTime', () => {
        it('should contains proper value', () => {
          // given
          const wrapper = shallow(<Trip {...props} />);

          const { value } = wrapper
            .find(CardContent)
            .at(0)
            .find(Grid)
            .at(1)
            .find(FormattedTime)
            .props();

          // then
          const expectedValue = new Date(props.departure_time);
          expect(value).toEqual(expectedValue);
        });
      });
    });
    describe('Duration', () => {
      it('should have proper className props', () => {
        // given
        const wrapper = shallow(<Trip {...props} />);

        const { className } = wrapper
          .find(CardContent)
          .at(0)
          .find(Grid)
          .at(2)
          .props();

        expect(className).toBe(props.classes.duration);
      });
      describe('Typography', () => {
        it('should have proper className props', () => {
          // given
          const wrapper = shallow(<Trip {...props} />);

          const { className } = wrapper
            .find(CardContent)
            .at(0)
            .find(Grid)
            .at(2)
            .find(Typography)
            .props();

          // then
          expect(className).toBe(props.classes.duration);
        });
        it('should contains proper hours message', () => {
          // given
          const wrapper = shallow(<Trip {...props} />);
          const hours = 8;

          const formattedMessageProps = wrapper
            .find(CardContent)
            .find(Typography)
            .find(FormattedMessage)
            .find('.duration__hours')
            .props();

          // then
          expect(formattedMessageProps).toMatchObject({
            values: { amount: hours },
            ...messages.hours,
          });
        });
        it('should contains proper minutes message', () => {
          // given
          const wrapper = shallow(<Trip {...props} />);
          const minutes = 19;

          const formattedMessageProps = wrapper
            .find(CardContent)
            .find(Typography)
            .find(FormattedMessage)
            .find('.duration__minutes')
            .props();

          // then
          expect(formattedMessageProps).toMatchObject({
            values: { amount: minutes },
            ...messages.minutes,
          });
        });
      });
    });
    describe('Arrival', () => {
      it('should have proper className props', () => {
        // given
        const wrapper = shallow(<Trip {...props} />);

        const { className } = wrapper
          .find(CardContent)
          .at(0)
          .find(Grid)
          .at(3)
          .props();

        expect(className).toBe(props.classes.arrival);
      });
      describe('FormattedDate', () => {
        it('should contains proper value', () => {
          // given
          const wrapper = shallow(<Trip {...props} />);

          const { value } = wrapper
            .find(CardContent)
            .at(0)
            .find(Grid)
            .at(3)
            .find(FormattedDate)
            .props();

          // then
          const expectedValue = new Date(props.arrival_time);
          expect(value).toEqual(expectedValue);
        });
      });
      describe('FormattedTime', () => {
        it('should contains proper value', () => {
          // given
          const wrapper = shallow(<Trip {...props} />);

          const { value } = wrapper
            .find(CardContent)
            .at(0)
            .find(Grid)
            .at(3)
            .find(FormattedTime)
            .props();

          // then
          const expectedValue = new Date(props.arrival_time);
          expect(value).toEqual(expectedValue);
        });
      });
    });
  });
  describe('Collapse', () => {
    it('should container proper in props', () => {
      // given
      const wrapper = shallow(<Trip {...props} />);

      const inProps = wrapper.find(Collapse).props().in;

      expect(inProps).toBe(props.isExpanded);
    });
    describe('Departure', () => {
      it('should have proper className props', () => {
        // given
        const wrapper = shallow(<Trip {...props} />);

        const { className } = wrapper
          .find(Collapse)
          .find(Grid)
          .at(1)
          .props();

        expect(className).toBe(props.classes.departure);
      });
      // given
      const wrapper = shallow(<Trip {...props} />);
      const minutes = 19;

      const formattedMessageProps = wrapper
        .find(CardContent)
        .find(Typography)
        .find(FormattedMessage)
        .find('.duration__minutes')
        .props();

      // then
      expect(formattedMessageProps).toMatchObject({
        values: { amount: minutes },
        ...messages.minutes,
      });
    });
  });
});
