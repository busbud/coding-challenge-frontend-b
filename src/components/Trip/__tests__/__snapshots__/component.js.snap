// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component | Jumbotron | component render should match snapshot 1`] = `
<WithStyles(GridWrapper)
  item={true}
  md={6}
  xs={12}
>
  <Card
    raised={false}
  >
    <WithStyles(CardContent)>
      <WithStyles(GridWrapper)
        alignItems="center"
        container={true}
        justify="center"
      >
        <WithStyles(GridWrapper)
          className="Component-departure-274"
          item={true}
          md={5}
          xs={6}
        >
          <WithStyles(Typography)
            gutterBottom={true}
          >
            <FormattedDate
              day="2-digit"
              month="long"
              value={2018-08-07T00:01:00.000Z}
            />
          </WithStyles(Typography)>
          <WithStyles(Typography)
            variant="display1"
          >
            <FormattedTime
              value={2018-08-07T00:01:00.000Z}
            />
          </WithStyles(Typography)>
        </WithStyles(GridWrapper)>
        <Hidden
          implementation="js"
          lgDown={false}
          lgUp={false}
          mdDown={true}
          mdUp={false}
          smDown={false}
          smUp={false}
          xlDown={false}
          xlUp={false}
          xsDown={false}
          xsUp={false}
        >
          <WithStyles(GridWrapper)
            className="Component-duration-275"
            item={true}
            xs={2}
          >
            <pure(DirectionsBus)
              color="action"
            />
            <WithStyles(Typography)
              className="Component-duration-275"
            >
              <FormattedMessage
                className="duration__hours"
                defaultMessage="{amount}H "
                id="components.trip.hours"
                values={
                  Object {
                    "amount": 8,
                  }
                }
              />
              <FormattedMessage
                className="duration__minutes"
                defaultMessage="{amount}m"
                id="components.trip.minutes"
                values={
                  Object {
                    "amount": 19,
                  }
                }
              />
            </WithStyles(Typography)>
          </WithStyles(GridWrapper)>
        </Hidden>
        <WithStyles(GridWrapper)
          className="Component-arrival-276"
          item={true}
          md={5}
          xs={6}
        >
          <WithStyles(Typography)
            gutterBottom={true}
          >
            <FormattedDate
              day="2-digit"
              month="long"
              value={2018-08-07T08:20:00.000Z}
            />
          </WithStyles(Typography)>
          <WithStyles(Typography)
            variant="display1"
          >
            <FormattedTime
              value={2018-08-07T08:20:00.000Z}
            />
          </WithStyles(Typography)>
        </WithStyles(GridWrapper)>
      </WithStyles(GridWrapper)>
    </WithStyles(CardContent)>
    <WithStyles(Collapse)
      in={false}
      timeout="auto"
      unmountOnExit={true}
    >
      <WithStyles(CardContent)>
        <WithStyles(GridWrapper)
          alignItems="center"
          container={true}
          justify="space-between"
        >
          <WithStyles(GridWrapper)
            className="Component-departure-274"
            item={true}
            xs={5}
          >
            <WithStyles(Typography)
              variant="title"
            >
              Port Authority Bus Terminal
            </WithStyles(Typography)>
            <WithStyles(Typography)>
              619-623 8th Ave, New York, NY 10018, USA
            </WithStyles(Typography)>
          </WithStyles(GridWrapper)>
          <WithStyles(GridWrapper)
            className="Component-arrival-276"
            item={true}
            xs={5}
          >
            <WithStyles(Typography)
              variant="title"
            >
              Gare d'autocars de Montréal
            </WithStyles(Typography)>
            <WithStyles(Typography)>
              1717 Rue Berri, Montréal, QC H2L 4E9, Canada
            </WithStyles(Typography)>
          </WithStyles(GridWrapper)>
        </WithStyles(GridWrapper)>
      </WithStyles(CardContent)>
    </WithStyles(Collapse)>
    <WithStyles(CardActions)
      className="Component-actions-277"
      disableActionSpacing={true}
    >
      <WithStyles(Typography)
        variant="button"
      >
        <FormattedNumber
          currency="USD"
          style="currency"
          value={72}
        />
      </WithStyles(Typography)>
      <WithStyles(Button)
        color="primary"
        href="https://www.busbud.com/en/deeplink/dr5reg/f25dvk/N2Y5ODU1MTpjMWM4Yzg0Mg?outbound_date=2018-08-07&return_date&adults=1&children=0&seniors=0&child_ages=&senior_ages=&discount_code&utm_source={utm_source}&utm_campaign={utm_campaign}&utm_medium={utm_medium}"
        target="_blank"
      >
        <FormattedMessage
          defaultMessage="Book now"
          id="components.trip.action.booking"
          values={Object {}}
        />
      </WithStyles(Button)>
      <WithStyles(IconButton)
        aria-expanded={false}
        aria-label="Show more"
        className="Component-expand-278 "
        onClick={[Function]}
      >
        <pure(ExpandMore) />
      </WithStyles(IconButton)>
    </WithStyles(CardActions)>
  </Card>
</WithStyles(GridWrapper)>
`;
